File: upwork-intelligence/ingest-mcp-batch.mjs
#!/usr/bin/env node
/** node ingest-mcp-batch.mjs batch.json — array of { keyword, group, jobs } or { keyword, group, response } */
import fs from 'fs';
import path from 'path';

const ROOT = path.dirname(new URL(import.meta.url).pathname);
const batch = JSON.parse(fs.readFileSync(process.argv[2], 'utf8'));
const out = path.join(ROOT, 'run-results.jsonl');
const seen = new Set();
if (fs.existsSync(out)) {
  for (const line of fs.readFileSync(out, 'utf8').split('\n')) {
    if (!line.trim()) continue;
    seen.add(JSON.parse(line).keyword);
  }
}

function stripJob(j) {
  return {
    url: j.url?.split('?')[0],
    title: j.title,
    published_date: j.published_date,
    created_date: j.created_date,
    job_type: j.job_type,
    budget: j.budget,
    duration: j.duration,
    proposal_count: j.proposal_count,
    experience_level: j.experience_level,
    skills: j.skills,
    client: j.client,
    description_snippet: j.description_snippet?.slice?.(0, 120) ?? '',
  };
}

let added = 0;
for (const item of batch) {
  if (seen.has(item.keyword)) continue;
  const jobs = (item.jobs || item.response?.jobs || []).map(stripJob);
  fs.appendFileSync(out, JSON.stringify({ keyword: item.keyword, group: item.group, jobs }) + '\n');
  seen.add(item.keyword);
  added++;
}
console.log(JSON.stringify({ added, total: seen.size }));

File: upwork-intelligence/ingest_mcp_batch.py
#!/usr/bin/env python3
"""Merge slim MCP batch JSON into search_responses.json."""
import json
import sys
from pathlib import Path

ROOT = Path(__file__).parent
SEARCH = ROOT / "search_responses.json"
GROUPS = json.loads((ROOT / "keywords.json").read_text())["keywords"]


def slim_job(j: dict) -> dict:
    return {k: v for k, v in j.items() if k != "description_snippet"}


def slim_response(resp: dict) -> dict:
    out = {k: v for k, v in resp.items() if k not in ("description_snippet",)}
    if "jobs" in out:
        out["jobs"] = [slim_job(j) for j in out["jobs"]]
    return out


def main() -> None:
    batch = json.loads(sys.stdin.read())
    paired = json.loads(SEARCH.read_text()) if SEARCH.exists() else []
    index = {e["keyword"]: i for i, e in enumerate(paired)}
    for item in batch:
        kw = item["keyword"]
        entry = {
            "keyword": kw,
            "group": GROUPS.get(kw, "OTHER"),
            "response": slim_response(item["response"]),
        }
        if kw in index:
            paired[index[kw]] = entry
        else:
            paired.append(entry)
            index[kw] = len(paired) - 1
    SEARCH.write_text(json.dumps(paired, ensure_ascii=False))
    print(len(paired))


if __name__ == "__main__":
    main()
